src/Input/Tokenizer.php

```diff
@@ -0,0 +1,464 @@
+            return $tokens;
+        }
+
+        if ($this->isShortEqOptions($arg)) {
+            $parts = explode(Token::TOKEN_OPTION_EQ, $arg, 2);
+            $tokens[] = new Token(Token::TYPE_SHORT, $parts[0]);
+            if (!empty($parts[1])) {
+                $t = $this->tokenize($parts[1]);
+                array_push($tokens, ...$t);
+            }
+            return $tokens;
+        }
+
+        if ($this->isLongOption($arg)) {
+            $tokens[] = new Token(Token::TYPE_LONG, $arg);
+            return $tokens;
+        }
+
+        if ($this->isLongEqOption($arg)) {
+            $parts = explode(Token::TOKEN_OPTION_EQ, $arg, 2);
+            $tokens[] = new Token(Token::TYPE_LONG, $parts[0]);
+
+            if (!empty($parts[1])) {
+                $t = $this->tokenize($parts[1]);
+                array_push($tokens, ...$t);
+            }
+
+            return $tokens;
+        }
+
+        // Unclassified, treat as constant:
+        return [new Token(Token::TYPE_CONSTANT, $arg)];
+
+    }
+
+    /**
+     * Get the current token.
+     * For Iterator interface.
+     * 
+     * @return Token
+     */
+    public function current(): Token
+    {
+        return $this->tokens[$this->index];
+    }
+    
+    /**
+     * Get the next token.
+     * Without moving the pointer.
+     * 
+     * @param int $offset
+     * 
+     * @return Token|null
+     */
+    public function offset(int $offset): ?Token
+    {
+        if (isset($this->tokens[$this->index + $offset])) {
+            return $this->tokens[$this->index + $offset];
+        }
+        return null;
+    }
+
+    /**
+     * Move the pointer to the next token.
+     * For Iterator interface.
+     * 
+     * @return void
+     */
+    public function next(): void
+    {
+        $this->index++;
+    }
+
+    /**
+     * Get the current token index.
+     * For Iterator interface.
+     * 
+     * @return int
+     */
+    public function key(): int
+    {
+        return $this->index;
+    }
+
+    /**
+     * Check if the current token is valid.
+     * For Iterator interface.
+     * 
+     * @return bool
+     */
+    public function valid(): bool
+    {
+        return isset($this->tokens[$this->index]);
+    }
+
+    /**
+     * Rewind the pointer to the first token.
+     * For Iterator interface.
+     * 
+     * @return void
+     */
+    public function rewind(): void
+    {
+        $this->index = 0;
+    }
+
+    /**
+     * Get the current token if valid.
+     * 
+     * @return Token|null
+     */
+    public function validCurrent(): ?Token
+    {
+        if ($this->valid()) {
+            return $this->current();
+        }
+        return null;
+    }
+
+    /** 
+     * toString magic method.
+     * for debugging.
+     */
+    public function __toString()
+    {
+        $str = PHP_EOL;
+        foreach ($this->tokens as $token) {
+            $str .= " - ".$token . PHP_EOL;
+            if (!empty($token->nested)) {
```

Comment:
!empty check not needed unless nested can be null. but i think we can use empty array `[]` for nested as default value